HelgaShiryaeva/assignment2/BinaryTree.java

```diff
@@ -1,104 +1,166 @@
         }
+            printAncestorsRecursive(root, value);
+            System.out.println();
     }
 
-    private Node<T> printAncestorsRecursive(Node<T> node, Node<T> currentNode) throws OutOfTreeException {
+    private Node printAncestorsRecursive(Node currentNode, T value) {
         if (currentNode != null) {
-            if (!nodes.contains(currentNode)) {
-                throw new OutOfTreeException(currentNode);
+            if (currentNode.getValue().equals(value)) {
+                return currentNode;
             }
-            if (currentNode.getValue().equals(node.getValue())) {
+            Node left = printAncestorsRecursive(currentNode.getLeft(), value);
+            if (left != null) {
+                System.out.print(currentNode.getValue() + " ");
                 return currentNode;
             }
-            Node<T> left = printAncestorsRecursive(node, currentNode.getLeft());
-            Node<T> right = printAncestorsRecursive(node, currentNode.getRight());
-            if (left == null && right == null) {
+            Node right = printAncestorsRecursive(currentNode.getRight(), value);
+            if (right == null) {
                 return null;
             } else {
                 System.out.print(currentNode.getValue() + " ");
                 return currentNode;
             }
         }
-         return null;
+        return null;
     }
 
     /**
-     * Finds the lowest common ansestor of two nodes of a binary tree
-     * @return node which is the lowest common ancestor of two nodes
-     * @throws OutOfTreeException if either fisrt or second or both node(s) are not a part
+     * Finds the lowest common ancestor of two nodes of a binary tree
+     * @return value of the node which is the lowest common ancestor of two nodes
+     * @throws NoSuchElementException if either first or second or both node(s) are not a part
      * of a binary tree
      */
-    public Node<T> getLca(Node<T> first, Node<T> second) throws OutOfTreeException {
-        if (!nodes.contains(first) || !nodes.contains(second)) {
-            throw new OutOfTreeException("Node(s} is(are) not present in a binary tree");
+    public T getLca(T first, T second) {
+        if (!values.contains(first) || !values.contains(second)) {
+            throw new NoSuchElementException("Node(s} is(are) not present in a binary tree");
         }
-        return getLcaRecursive(root, first, second);
+        return getLcaRecursive(root, first, second).getValue();
     }
 
-    private Node<T> getLcaRecursive(Node<T> current, Node<T> first, Node<T> second) throws OutOfTreeException {
+    private Node getLcaRecursive(Node current, T first, T second) {
         if (current == null) {
             return null;
         }
-        if(!nodes.contains(current)) {
-            throw new OutOfTreeException(current);
-        }
-        if (current == first || current == second) {
+        if (current.getValue() == first || current.getValue() == second) {

```

Comment:
Comparison of values should be done with `equals()` rather than `==`. For example two `Integer`s can carry the same value but not be equal when compared with `==`.